// rdfjs-wrapper/dist/mapping/TermMapping.js
import { TermWrapper } from "../TermWrapper.js";
/*
 * Write to the RDF dataset
 * Transform JavaScript primitive types to Term Wrapper
 * Generic Term Wrapper factories
*/
export var TermMapping;
(function (TermMapping) {
    function dateToLiteral(value, dataset, factory) {
        return new TermWrapper(factory.literal(value.toISOString(), factory.namedNode("http://www.w3.org/2001/XMLSchema#date")), dataset, factory);
    }
    TermMapping.dateToLiteral = dateToLiteral;
    // TODO: Lang string dictionary value mapping
    function langStringToLiteral(value, dataset, factory) {
        // TODO: Fix bug in N3.JS function literal line 340 (test languageOrDataType === 'string')
        // Then re-establish proper factory method
        //return new TermWrapper(factory.literal(value.string, { language: value.lang } ), dataset, factory)
        return new TermWrapper(factory.literal(value.string, value.lang), dataset, factory);
    }
    TermMapping.langStringToLiteral = langStringToLiteral;
    function numberToLiteral(value, dataset, factory) {
        return new TermWrapper(factory.literal(value.toString(), factory.namedNode("http://www.w3.org/2001/XMLSchema#double")), dataset, factory);
    }
    TermMapping.numberToLiteral = numberToLiteral;
    function stringToBlankNode(value, dataset, factory) {
        return new TermWrapper(factory.blankNode(value), dataset, factory);
    }
    TermMapping.stringToBlankNode = stringToBlankNode;
    function stringToIri(value, dataset, factory) {
        return new TermWrapper(factory.namedNode(value), dataset, factory);
    }
    TermMapping.stringToIri = stringToIri;
    function stringToLiteral(value, dataset, factory) {
        return new TermWrapper(factory.literal(value), dataset, factory);
    }
    TermMapping.stringToLiteral = stringToLiteral;
    function asIs(value, dataset, factory) {
        return new TermWrapper(value, dataset, factory);
    }
    TermMapping.asIs = asIs;
    function identity(value) {
        return value;
    }
    TermMapping.identity = identity;
})(TermMapping || (TermMapping = {}));
