// rdfjs-wrapper/dist/mapping/ValueMapping.js
/*
 * Read from the RDF dataset
 * Transform RDF Terms to JavaScript primitive types
*/
export var ValueMapping;
(function (ValueMapping) {
    function blankNodeToString(termWrapper) {
        // TODO: Throw typed error
        if (termWrapper.term.termType != "BlankNode") {
            throw Error(`Value mapping expected blank node but got ${termWrapper.term.termType}`);
        }
        return iriOrBlankNodeToString(termWrapper);
    }
    ValueMapping.blankNodeToString = blankNodeToString;
    function literalToDate(termWrapper) {
        // TODO: Check term type
        return new Date(termWrapper.term.value);
    }
    ValueMapping.literalToDate = literalToDate;
    // TODO: Lang string dictionary value mapping
    function literalToLangString(termWrapper) {
        // TODO: Check term type
        const x = termWrapper.term;
        return { lang: x.language, string: x.value };
    }
    ValueMapping.literalToLangString = literalToLangString;
    function literalToNumber(termWrapper) {
        // TODO: Check term type
        return Number(termWrapper.term.value);
    }
    ValueMapping.literalToNumber = literalToNumber;
    function literalToString(termWrapper) {
        // TODO: Check term type
        return termWrapper.term.value;
    }
    ValueMapping.literalToString = literalToString;
    function iriToString(termWrapper) {
        // TODO: Check term type
        return iriOrBlankNodeToString(termWrapper);
    }
    ValueMapping.iriToString = iriToString;
    function iriOrBlankNodeToString(termWrapper) {
        return termWrapper.term.value;
    }
    ValueMapping.iriOrBlankNodeToString = iriOrBlankNodeToString;
    function asIs(termWrapper) {
        return termWrapper.term;
    }
    ValueMapping.asIs = asIs;
})(ValueMapping || (ValueMapping = {}));
