// rdfjs-wrapper/dist/DatasetWrapper.js
import { RDF } from "./vocabulary/RDF.js";
class DatasetCoreBase {
    dataset;
    factory;
    constructor(dataset, factory) {
        this.dataset = dataset;
        this.factory = factory;
    }
    get size() {
        return this.dataset.size;
    }
    [Symbol.iterator]() {
        return this.dataset[Symbol.iterator]();
    }
    add(quad) {
        this.dataset.add(quad);
        return this;
    }
    delete(quad) {
        this.dataset.delete(quad);
        return this;
    }
    has(quad) {
        return this.dataset.has(quad);
    }
    match(subject, predicate, object, graph) {
        return this.dataset.match(subject, predicate, object, graph);
    }
}
export class DatasetWrapper extends DatasetCoreBase {
    *subjectsOf(predicate, termWrapper) {
        for (const q of this.matchSubjectsOf(termWrapper, predicate)) {
            yield q;
        }
    }
    *matchSubjectsOf(termWrapper, predicate, object, graph) {
        const p = predicate ? this.factory.namedNode(predicate) : undefined;
        const o = object ? this.factory.namedNode(object) : undefined;
        const g = graph ? this.factory.namedNode(graph) : undefined;
        for (const q of this.match(undefined, p, o, g)) {
            yield new termWrapper(q.subject, this, this.factory);
        }
    }
    *objectsOf(predicate, termWrapper) {
        for (const q of this.matchObjectsOf(termWrapper, undefined, predicate)) {
            yield q;
        }
    }
    *matchObjectsOf(termWrapper, subject, predicate, graph) {
        const s = subject ? this.factory.namedNode(subject) : undefined;
        const p = predicate ? this.factory.namedNode(predicate) : undefined;
        const g = graph ? this.factory.namedNode(graph) : undefined;
        for (const q of this.match(s, p, undefined, g)) {
            yield new termWrapper(q.object, this, this.factory);
        }
    }
    *instancesOf(type, constructor) {
        for (const q of this.matchSubjectsOf(constructor, RDF.type, type)) {
            yield q;
        }
    }
}
