// rdfjs-wrapper/dist/ListItem.js
import { TermWrapper } from "./TermWrapper.js";
import { ValueMapping } from "./mapping/ValueMapping.js";
import { TermMapping } from "./mapping/TermMapping.js";
import { RDF } from "./vocabulary/RDF.js";
export class ListItem extends TermWrapper {
    valueMapping;
    termMapping;
    constructor(term, dataset, factory, valueMapping, termMapping) {
        super(term, dataset, factory);
        this.valueMapping = valueMapping;
        this.termMapping = termMapping;
    }
    get firstRaw() {
        return this.singularNullable(RDF.first, ValueMapping.asIs);
    }
    set firstRaw(value) {
        this.overwriteNullable(RDF.first, value, TermMapping.asIs);
    }
    get restRaw() {
        return this.singularNullable(RDF.rest, ValueMapping.asIs);
    }
    set restRaw(value) {
        this.overwriteNullable(RDF.rest, value, TermMapping.asIs);
    }
    get isListItem() {
        return this.firstRaw !== undefined && this.restRaw !== undefined;
    }
    get isNil() {
        return this.term.equals(this.factory.namedNode(RDF.nil));
    }
    get first() {
        return this.singular(RDF.first, this.valueMapping);
    }
    set first(value) {
        this.overwrite(RDF.first, value, this.termMapping);
    }
    get rest() {
        return this.singular(RDF.rest, w => new ListItem(w.term, w.dataset, w.factory, this.valueMapping, this.termMapping));
    }
    set rest(value) {
        this.overwrite(RDF.rest, value, TermMapping.identity);
    }
    pop() {
        try {
            return this.first;
        }
        finally {
            this.firstRaw = undefined;
            this.restRaw = this.factory.namedNode(RDF.nil);
        }
    }
    *items() {
        if (this.firstRaw === undefined) {
            return;
        }
        yield this;
        for (const more of this.rest.items()) {
            yield more;
        }
    }
}
